### PR TITLE
USB: Move to Qt (with *significant* refactoring)

### LINKED ISSUE #4041: [Feature Request] USB - Sega Seamic Custom Controller Setup Screen
**Description**
The sega seamic is a unique controller that doesn't exactly follow a standard ps2 controller and it has a built in microphone. The current setup screen was designed for steering wheels

Controller images and explainations are here
https://github.com/jackun/USBqemu-wheel/issues/44

### PR DESCRIPTION
**Pull request by stenzek**
### Description of Changes

Still a work in progress. WX no longer has USB, it's silly to keep two versions in the tree when it won't be around much longer, and I changed the interface anyway.

New USB has:
 - Actually working save states.
 - Mac compatibility (less eyetoy).
 - A consistent UI with the rest of the application.
 - Correct binding listing based on the controller type.
 - GunCon2 support.

TODO:

- [x] Fix incorrect lightgun positioning when there's borders around the display.
- [x] Eyetoy/usb mic.
- [x] Add bindings for remaining pad types.
- [x] Attempt to add force feedback via SDL (but I can't test it..)
- [x] Fix USB keyboards.
- ~Add USB device configuration to Fullscreen UI.~ (deferred for after merge)
- [x] Add USB device bindings to input overlay.
- ~Add multiple mouse support for USB mouse and lightgun.~ (deferred for after merge)
- [x] Fix saving state for USB printer.
- [x] Support binding no device to USB mic.
- [x] Access e.g. HCCA directly instead of copying (and skip with no devices connected).
- [x] Add OSD message for printer files being saved.

### Rationale behind Changes

Be gone wx you disgusting horribly written overengineered turd sandwich. The USB code I ripped out and had to rewrite was just as bad, full of undefined behaviour and memory leaks.

### Suggested Testing Steps

Test new functionality.

Closes #4041.

### CODE REVIEW

**Comment by seta-san** [comment]
as soon as it's ready i'll be here to go nuts testing this. one note on the microphone; this https://github.com/PCSX2/pcsx2/issues/4167 has the descriptors for the konami microphone for japanese karaoke revolution games and i would like to see if it can be more universally used

**Comment by refractionpcsx2** [comment]
Well once Sten has finished this, you're free to implement them :)

**Comment by TheComputerGuy96** [comment]
We do a bit of trolling :trollface: 

EyeToy Play game running inside the Qt build of PCSX2 on Linux (with Xbox One S controller shown in the image)

**Comment by refractionpcsx2** [comment]
Is that an xbox??

**Comment by TheComputerGuy96** [comment]
@refractionpcsx2 I showed a Xbox One S wireless controller in the camera feed

**Comment by TheComputerGuy96** [comment]
I recently tried this at my kid's birthday and it was fine :frog:

**Comment by seta-san** [comment]
eyetoy crashes immediately. the console log says it was unable to open the device. singstar boots and gets in game but no actual audio from a device makes it into the game.

**Comment by refractionpcsx2** [comment]
> eyetoy crashes immediately. the console log says it was unable to open the device. singstar boots and gets in game but no actual audio from a device makes it into the game.

Can you provide backtraces/.dmp files, please.

**Comment by seta-san** [comment]
[crash-2022-12-14-12-40-09-159.zip](https://github.com/PCSX2/pcsx2/files/10230718/crash-2022-12-14-12-40-09-159.zip)

File: pcsx2/USB/USB.cpp [hunk 1]
```diff
@@ -20,286 +20,174 @@
...
-		device = devProxy->CreateDevice(port);
-	else
-		Console.WriteLn(Color_Red, "Device %d: Unknown device type", 1 - port);
-
-	if (!device)
+	if (dev->klass.open && dev->klass.open(dev) != 0)
 	{
+		Console.Error("Failed to open USB device in port %u (%s)", port + 1, proxy->Name());
+		if (dev->klass.unrealize)
+			s_usb_device[port]->klass.unrealize(dev);
```
Comment (refractionpcsx2) [inline pcsx2/USB/USB.cpp:98]: 100% of errors (2 reports) are caused by this line. I presume you meant dev->klass.unrealize(dev);? s_usb_device is null for both ports and isn't set until after this.

**Comment by seta-san** [comment]
Camera works. Microphone games still aren’t getting their input.

**Comment by MarcoEstevez** [comment]
I'm not sure whether Force Feedback works, on the old GUI there was an option to test it, I'm missing it on the new GUI

image

**Comment by stenzek** [comment]
No, there's not going to be a test option in the new design. The UI is mostly unaware of the inner workings of USB, including FFB.

**Comment by MarcoEstevez** [comment]
Ok what about the other 4 options, will be eventually added

**Comment by stenzek** [comment]
> Ok what about the other 4 options, will be eventually added

What's there now is what you're getting. Rewriting all this stuff has been mental torture for me, especially since it's not something I care about in the slightest apart from lightguns. If someone wants to add more extensive options after it's merged, they're free to do so.

**Comment by MarcoEstevez** [comment]
> What's there now is what you're getting. Rewriting all this stuff has been mental torture for me, especially since it's not something I care about in the slightest apart from lightguns. If someone wants to add more extensive options after it's merged, they're free to do so.

@stenzek thanks for your quick answer, and also for all the great work you have done, I hope someone else picks it up and eventually add all the awesome features that will be left over

**Comment by stenzek** [comment]
All done. There's some stuff I listed above which I'm going to do post-merge, but definitely taking a break from USB for a while, so over it.

I've added SDL joystick handling to the input source, so in theory wheels should all work, including FFB, but again, no way of testing it.

**Review by lightningterror — COMMENTED** [review]
I see some new mac specific warnings in deviceproxy.h

https://github.com/PCSX2/pcsx2/actions/runs/3714569305/jobs/6298642652#step:11:973

**Comment by seta-san** [comment]
switching to singstar while in game can cause a crash
[crash-2022-12-16-13-27-50-769.zip](https://github.com/PCSX2/pcsx2/files/10248465/crash-2022-12-16-13-27-50-769.zip)

**Comment by refractionpcsx2** [comment]
> switching to singstar while in game can cause a crash [crash-2022-12-16-13-27-50-769.zip](https://github.com/PCSX2/pcsx2/files/10248465/crash-2022-12-16-13-27-50-769.zip)

What's in your singstar config? is it blank because you just switched to it while the game is running?

it's a bit weird cos the device name is blank, but somehow it got passed the "is blank" check... :/

**Comment by seta-san** [comment]
so it happens when i switch from a logitech microphone to the singstar...

**Comment by refractionpcsx2** [comment]
okay but is anything configured for singstar?

**Comment by seta-san** [comment]
yes. i think i found the problem. there was nothing configured for the logitech mic

**Comment by seta-san** [comment]
so yeah. if you switch from logitech mic with input device set to not connected and switch it to singstar it crashes even if singstar is configured

**Comment by seta-san** [comment]
something else i noticed is that if you switch to the Logitech microphone the game doesn't recognize that the singstar device was disconnected.... but it does if i switch it to the eyetoy or something else non-microphone related.

**Comment by seta-san** [comment]
seaman 2 doesn't recognize the seamic controller plugged in

**Comment by Mrlinkwii** [comment]
@seta-san  it dose 
image

**Comment by seta-san** [comment]
i think my problem is that i tried to bind it to the same controller as the normal pad

**Comment by stenzek** [comment]
> something else i noticed is that if you switch to the Logitech microphone the game doesn't recognize that the singstar device was disconnected.... but it does if i switch it to the eyetoy or something else non-microphone related.

No, worked fine for me. The replug logic doesn't care what kind of device it is, they all behave the same.

Crash when quickly switching between mic types should be fixed though.

File: pcsx2/USB/USB.cpp [hunk 2]
```diff
@@ -332,237 +196,311 @@ void USBwrite16(u32 addr, u16 value)
...
+			s32 state_devtype;
+			u32 state_devsubtype;
+			u32 state_size;
+			sw.Do(&state_devtype);
+			sw.Do(&state_devsubtype);
+			sw.Do(&state_size);
+
+			// this is *assuming* the config is correct... there's no reason it shouldn't be.
+			if (sw.HasError() ||
+				EmuConfig.USB.Ports[port].DeviceType != state_devtype ||
```
Comment (lightningterror) [inline pcsx2/USB/USB.cpp:410]: Here as well
https://github.com/PCSX2/pcsx2/actions/runs/3720548584/jobs/6310096432#step:11:960

**Comment by ROBYER1** [comment]
This is great stuff, does this PR carry across to the Retroarch core? I started work on this a while ago too but got stuck at fathoming out USB support.

Was interested in seeing if this could be hooked up to the Lightgun support in Retroarch so that PCSX2 Core can run in Emu VR to play PS2 lightgun games in VR 🥳

**Comment by refractionpcsx2** [comment]
We're nothing to do with retroarch. I don't this will ever appear there

**Comment by stenzek** [comment]
That hack of a thing really shouldn't be called PCSX2 anymore, we've rewritten so much in the last couple of years that it's nothing alike, not in performance, features, nor compatibility.

Besides, I don't understand why you're asking us that question. Shouldn't you be asking the "maintainers" of said core?

**Comment by ROBYER1** [comment]
Hey, I get that I've probably trodden on someone's toes here, I'm just a hobbyist developer interested in using PS2 games in VR for the lightgun games which happens through EmuVR who only support Retroarch.

I actually couldn't find the GitHub repo for their PCSX2 core although I'm on my phone atm as I'm out and about. I'll have a look later where I can raise that with them, if you do happen to have a link feel free to send me it. I didn't realise the PCSX2 Retroarch core was frowned upon by the PCSX2 team 😂

**Comment by TheComputerGuy96** [comment]
@ROBYER1 I actually found it on my phone: https://github.com/libretro/pcsx2

**Comment by ROBYER1** [comment]
> Besides, I don't understand why you're asking us that question. Shouldn't you be asking the "maintainers" of said core?

Why does the PCSX2 Retroarch core share the same name if it is so different and clearly frowned upon by the developers of the main PCSX2 emulator? Smh.

**Comment by legend800** [comment]
Stop spamming the thread with unrelated conversation. Take it to the forums

**Comment by F0bes** [comment]
> Why does the PCSX2 Retroarch core share the same name if it is so different and clearly frowned upon by the developers of the main PCSX2 emulator? Smh.

Unfortunately it is out of the (actual) PCSX2 team's control.



> Stop spamming the thread with unrelated conversation. Take it to the forums

Who are you?

**Pull request merged**